### PR TITLE
Add tests for task model, add and delete lesson commands and command parsers

### PR DESCRIPTION
**Pull request by laxus2308**
- Add in Task util classes for testing
- Add task test
- Add task description test
- Add task deadline test
- Add student matric number test
- Refactor `DeleteStudentCommandTest`
- Add `DeleteTaskCommandTest`
- Add AddLessonCommandTest and AddLessonCommandParserTest for mc, studio, consult
- Update parserutil test

### CODE REVIEW

File: src/test/java/jarvis/model/student/StudentMatricNumTest.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        assertFalse(MatricNum.isValidMatricNum("")); // empty string
+        assertFalse(MatricNum.isValidMatricNum(" ")); // spaces only
+        assertFalse(MatricNum.isValidMatricNum("B1234567U")); // not start with A
+        assertFalse(MatricNum.isValidMatricNum("a1234567U")); // starts with small letter A
+        assertFalse(MatricNum.isValidMatricNum("A1234567u")); // ends with small letter
+        assertFalse(MatricNum.isValidMatricNum("A123456U")); // 6 digits
+
+        // valid matric number
+        assertTrue(MatricNum.isValidMatricNum("A1234567U")); // alphabets only
+        assertTrue(MatricNum.isValidMatricNum("A1234567A")); // numbers only
```
Comment (neosunhan) [inline src/test/java/jarvis/model/student/StudentMatricNumTest.java:39]: What do these comments mean?

File: src/test/java/jarvis/model/student/StudentMatricNumTest.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    @Test
+    public void equals() {
+
+    }
```
Comment (neosunhan) [inline src/test/java/jarvis/model/student/StudentMatricNumTest.java:45]: I think you forgot to test this?

File: src/test/java/jarvis/logic/commands/AddMasteryCheckCommandTest.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+        ModelStubWithLesson modelStub = new ModelStubWithLesson(validMasteryCheck);
+
+        assertThrows(CommandException.class, AddMasteryCheckCommand.MESSAGE_DUPLICATE_MASTERY_CHECK, () ->
+                addMasteryCheckCommand.execute(modelStub));
+    }
+
+    @Test
+    public void execute_clashTimePeriod_throwsCommandException() {
+        //valid masteryCheck with diff desc but same time period
```
Comment (fongyj) [inline src/test/java/jarvis/logic/commands/AddMasteryCheckCommandTest.java:69]:
```suggestion
        // valid masteryCheck with diff desc but same time period
```

File: src/test/java/jarvis/logic/commands/AddStudioCommandTest.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+        ModelStubWithLesson modelStub = new ModelStubWithLesson(validStudio);
+
+        assertThrows(CommandException.class,
+                AddStudioCommand.MESSAGE_DUPLICATE_STUDIO, () -> addStudioCommand.execute(modelStub));
+    }
+
+    @Test
+    public void execute_clashTimePeriod_throwsCommandException() {
+        //valid studio with diff desc but same time period
```
Comment (fongyj) [inline src/test/java/jarvis/logic/commands/AddStudioCommandTest.java:56]:
```suggestion
        // valid studio with diff desc but same time period
```

File: src/test/java/jarvis/logic/commands/AddConsultCommandTest.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+        ModelStubWithLesson modelStub = new ModelStubWithLesson(validConsult);
+
+        assertThrows(CommandException.class,
+                AddConsultCommand.MESSAGE_DUPLICATE_CONSULT, () -> addConsultCommand.execute(modelStub));
+    }
+
+    @Test
+    public void execute_clashTimePeriod_throwsCommandException() {
+        //valid consult with diff desc but same time period
```
Comment (fongyj) [inline src/test/java/jarvis/logic/commands/AddConsultCommandTest.java:69]:
```suggestion
        // valid consult with diff desc but same time period
```

File: src/test/java/jarvis/logic/commands/DeleteTaskCommandTest.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    //        showTaskAtIndex(model, INDEX_FIRST_TASK);
+    //
+    //        Index outOfBoundIndex = INDEX_SECOND_TASK;
+    //        // ensures that outOfBoundIndex is still in bounds of task book list
+    //        assertTrue(outOfBoundIndex.getZeroBased() < model.getTaskBook().getTaskList().size());
+    //
+    //        DeleteTaskCommand deleteTaskCommand = new DeleteTaskCommand(outOfBoundIndex);
+    //
+    //        assertCommandFailure(deleteTaskCommand, model, Messages.MESSAGE_INVALID_TASK_DISPLAYED_INDEX);
+    //    }
```
Comment (fongyj) [inline src/test/java/jarvis/logic/commands/DeleteTaskCommandTest.java:76]: did you forget to delete this

File: src/test/java/jarvis/model/StudentMatricNumTest.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        assertTrue(MatricNum.isValidMatricNum("A1234567U"));
+        assertTrue(MatricNum.isValidMatricNum("A1234567A"));
+    }
+
+    @Test
+    public void equals() {
+        MatricNum matricNum1 = new MatricNum(VALID_MATRIC_NUM_AMY);
+        MatricNum matricNum2 = new MatricNum(VALID_MATRIC_NUM_BOB);
+
+        //same values -> returns true
```
Comment (fongyj) [inline src/test/java/jarvis/model/StudentMatricNumTest.java:47]:
```suggestion
        // same values -> returns true
```

File: src/test/java/jarvis/model/StudentMatricNumTest.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        MatricNum matricNum1Copy = new MatricNum(VALID_MATRIC_NUM_AMY);
+        assertTrue(matricNum1.equals(matricNum1Copy));
+
+        //same object -> returns true
```
Comment (fongyj) [inline src/test/java/jarvis/model/StudentMatricNumTest.java:51]:
```suggestion
        // same object -> returns true
```

File: src/test/java/jarvis/model/StudentMatricNumTest.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        assertTrue(matricNum1.equals(matricNum1));
+
+        // null -> returns false
+        assertFalse(matricNum1.equals(null));
+
+        // different type -> returns false
+        assertFalse(matricNum1.equals(5));
+
+        //different values -> returns false
```
Comment (fongyj) [inline src/test/java/jarvis/model/StudentMatricNumTest.java:60]:
```suggestion
        // different values -> returns false
```

File: src/test/java/jarvis/model/TaskDescTest.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        assertTrue(desc1.equals(desc1));
+
+        // null -> returns false
+        assertFalse(desc1.equals(null));
+
+        // different type -> returns false
+        assertFalse(desc1.equals(5));
+
+        //different values -> returns false
+        assertFalse(desc1.equals(desc2));
```
Comment (fongyj) [inline src/test/java/jarvis/model/TaskDescTest.java:56]:
```suggestion
        // same values -> returns true
        TaskDesc desc1Copy = new TaskDesc("desc1");
        assertTrue(desc1.equals(desc1Copy));

        // same object -> returns true
        assertTrue(desc1.equals(desc1));

        // null -> returns false
        assertFalse(desc1.equals(null));

        // different type -> returns false
        assertFalse(desc1.equals(5));

        // different values -> returns false
        assertFalse(desc1.equals(desc2));
```

File: src/test/java/jarvis/model/TaskDeadlineTest.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+        assertTrue(deadline1.equals(deadline1));
+
+        // null -> returns false
+        assertFalse(deadline1.equals(null));
+
+        // different type -> returns false
+        assertFalse(deadline1.equals(5));
+
+        //different values -> returns false
+        assertFalse(deadline1.equals(deadline2));
```
Comment (fongyj) [inline src/test/java/jarvis/model/TaskDeadlineTest.java:46]:
```suggestion
        // same values -> returns true
        TaskDeadline deadline1Copy = new TaskDeadline(date1);
        assertTrue(deadline1.equals(deadline1Copy));

        // same object -> returns true
        assertTrue(deadline1.equals(deadline1));

        // null -> returns false
        assertFalse(deadline1.equals(null));

        // different type -> returns false
        assertFalse(deadline1.equals(5));

        // different values -> returns false
        assertFalse(deadline1.equals(deadline2));
```

File: src/test/java/jarvis/model/TaskTest.java [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    @Test
+    public void getDeadlineString() {
+        //no deadline
+        Task taskWithoutDeadline = new TaskBuilder().withoutDeadline().build();
+        assertEquals("", taskWithoutDeadline.getDeadlineString());
+
+        //has deadline
+        Task taskWithDeadline = new TaskBuilder().withDeadline(VALID_TASK_DEADLINE).build();
+        String expectedDeadline = VALID_TASK_DEADLINE.format(DateTimeFormatter.ofPattern("MMM-dd-yyyy"));
+        assertEquals(expectedDeadline, taskWithDeadline.getDeadlineString());
```
Comment (fongyj) [inline src/test/java/jarvis/model/TaskTest.java:88]:
```suggestion
    public void getDeadline() {
        // no deadline
        Task taskWithoutDeadline = new TaskBuilder().withoutDeadline().build();
        assertEquals(null, taskWithoutDeadline.getDeadline());

        // correct deadline
        Task taskWithDeadline = new TaskBuilder().withDeadline(VALID_TASK_DEADLINE).build();
        assertEquals(VALID_TASK_DEADLINE, taskWithDeadline.getDeadline());
    }

    @Test
    public void getDeadlineString() {
        // no deadline
        Task taskWithoutDeadline = new TaskBuilder().withoutDeadline().build();
        assertEquals("", taskWithoutDeadline.getDeadlineString());

        // has deadline
        Task taskWithDeadline = new TaskBuilder().withDeadline(VALID_TASK_DEADLINE).build();
        String expectedDeadline = VALID_TASK_DEADLINE.format(DateTimeFormatter.ofPattern("MMM-dd-yyyy"));
        assertEquals(expectedDeadline, taskWithDeadline.getDeadlineString());
```

File: src/test/java/jarvis/model/TaskTest.java [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+
+        newTask.markAsDone(); // task that is already done remains done
+        assertTrue(newTask.isDone());
+    }
+
+    @Test
+    public void markAsNotDone_isTaskDone_returnsFalse() {
+        Task newTask = new TaskBuilder().build();
+        newTask.markAsNotDone();
+        assertFalse(newTask.isDone()); //task that is not done remains not done
```
Comment (fongyj) [inline src/test/java/jarvis/model/TaskTest.java:60]:
```suggestion
        assertFalse(newTask.isDone()); // task that is not done remains not done
```

**Review by fongyj — CHANGES_REQUESTED** [review]

**Review by fongyj — APPROVED** [review]
LGTM

**Pull request merged**